library/src/include/lapack_device_functions.hpp

```diff
@@ -1111,4 +1111,96 @@ ROCSOLVER_KERNEL void axpy_kernel(const rocblas_int n,
     }
 }
 
+template <typename S, typename T, typename I>
+__global__ static void
+    rot_kernel(I const n, T* const x, I const incx, T* const y, I const incy, S const c, S const s)
+{
+    if(n <= 0)
+        return;
+
+    I const i_start = hipThreadIdx_x + hipBlockIdx_x * hipBlockDim_x;
+    I const i_inc = hipBlockDim_x * hipGridDim_x;
+
+    if((incx == 1) && (incy == 1))
+    {
+        // ------------
+        // special case
+        // ------------
+        for(I i = i_start; i < n; i += i_inc)
+        {
+            auto const temp = c * x[i] + s * y[i];
+            y[i] = c * y[i] - s * x[i];
+            x[i] = temp;
+        }
+    }
+    else
+    {
+        // ---------------------------
+        // code for unequal increments
+        // ---------------------------
+
+        for(auto i = i_start; i < n; i += i_inc)
+        {
+            auto const ix = 0 + i * static_cast<int64_t>(incx);
+            auto const iy = 0 + i * static_cast<int64_t>(incy);
+            auto const temp = c * x[ix] + s * y[iy];
+            y[iy] = c * y[iy] - s * x[ix];
+            x[ix] = temp;
+        }
+    }
+}
+
+template <typename S, typename T, typename I>
+static void
+    rot_template(I const n, T* x, I const incx, T* y, I const incy, S const c, S const s, hipStream_t stream)
+{
+    auto nthreads = warpSize * 2;
+    auto nblocks = (n - 1) / nthreads + 1;
+
+    hipLaunchKernelGGL((rot_kernel<S, T, I>), dim3(nblocks, 1, 1), dim3(nthreads, 1, 1), 0, stream,
+                       n, x, incx, y, incy, c, s);
+}
```

Comment:
I think we can leave it there for this release. Ultimately I would like to remove these methods and integrate their functionality directly into the main hybrid code, similar to how other functions launch their kernels. I'd like to query the wrap size from the device, since I think we established that the `warpSize` variable does not actually reflect the real warp size of the device, and doing that from the main hybrid code would be cleanest.

Reply:
Sure, I am ok with leaving the code there for clarity -- I won't add those comments related to the placement of host code to the new PR.

Reply:
mmm, good observations here!!
I did not think about this when I rearranged the code. 

The `_template` methods are just the kernel launchers, and the kernels themselves are supposed to work with any launch configuration (one of the advantages of the for-loop style). So, it makes more sense to me that the kernel launchers are in the function that launches them because the launch configuration is meaningful to that function. 

I would like to move them. When we revisit the code this may change, but at the current state, it makes more sense. 

I made the change, let me know what you think.